core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/ruin/RuinCustomPhase.java

```diff
@@ -0,0 +1,92 @@
+
+import org.optaplanner.core.api.domain.solution.PlanningSolution;
+import org.optaplanner.core.api.score.Score;
+import org.optaplanner.core.impl.heuristic.selector.move.MoveSelector;
+import org.optaplanner.core.impl.phase.custom.CustomPhase;
+import org.optaplanner.core.impl.phase.custom.CustomPhaseCommand;
+import org.optaplanner.core.impl.phase.custom.DefaultCustomPhase;
+import org.optaplanner.core.impl.phase.custom.scope.CustomPhaseScope;
+import org.optaplanner.core.impl.phase.custom.scope.CustomStepScope;
+import org.optaplanner.core.impl.solver.scope.SolverScope;
+import org.optaplanner.core.impl.solver.termination.Termination;
+
+/**
+ * Ruin implementation of {@link CustomPhase}.
+ *
+ * @param <Solution_> the solution type, the class with the {@link PlanningSolution} annotation
+ */
+public class RuinCustomPhase<Solution_, Score_ extends Score<Score_>> extends DefaultCustomPhase<Solution_> {
+
+    private Solution_ bestSolutionBeforeRuin;
+    private Score_ bestScoreBeforeRuin;
+    private final MoveSelector<Solution_> moveSelector;
+
+    protected RuinCustomPhase(Builder<Solution_> builder) {
+        super(builder);
+        this.moveSelector = builder.moveSelector;
+    }
+
+    // ************************************************************************
+    // Worker methods
+    // ************************************************************************
+
+    @Override
+    public void solvingStarted(SolverScope<Solution_> solverScope) {
+        super.solvingStarted(solverScope);
+        moveSelector.solvingStarted(solverScope);
+    }
+
+    public void phaseStarted(CustomPhaseScope<Solution_> phaseScope) {
+        super.phaseStarted(phaseScope);
+        moveSelector.phaseStarted(phaseScope);
+        bestSolutionBeforeRuin = phaseScope.getScoreDirector().cloneWorkingSolution();
+        bestScoreBeforeRuin = (Score_) phaseScope.getScoreDirector().calculateScore();
+    }
+
+    public void stepStarted(CustomStepScope<Solution_> stepScope) {
+        super.stepStarted(stepScope);
+        moveSelector.stepStarted(stepScope);
+    }
+
+    public void stepEnded(CustomStepScope<Solution_> stepScope) {
+        super.stepEnded(stepScope);
+        moveSelector.stepEnded(stepScope);
+    }
+
+    public void phaseEnded(CustomPhaseScope<Solution_> phaseScope) {
+        super.phaseEnded(phaseScope);
+        moveSelector.phaseStarted(phaseScope);
+    }
+
+    @Override
+    public void solvingEnded(SolverScope<Solution_> solverScope) {
+        if (bestScoreBeforeRuin.compareTo((Score_) solverScope.getBestScore()) > 0) {
+            solverScope.setBestSolution(bestSolutionBeforeRuin);
+            solverScope.setBestScore(bestScoreBeforeRuin);
+            solverScope.setBestSolutionTimeMillis(System.currentTimeMillis());
+            solverScope.getScoreDirector().setWorkingSolution(bestSolutionBeforeRuin);
+            solver.getBestSolutionRecaller().updateBestSolutionAndFire(solverScope);
+        }
```

Comment:
Ok, I understand the use case now, thank you for that. 
Still, it seems that we are solving a symptom, and not the root cause.

> If we allow destructive moves that worsen solution and set that as current best solution

Why would we allow to set something as best, if it's not in fact better than the best we already have? That destroys the concept of the best solution - which is always the best solution that the solver reached so far, regardless of what happened in the meantime.

If there is code that sets the best solution to a *worse* solution than we had before, I'd argue that this is the problem we have to fix. New best solution must only ever be better than the old best solution.

Reply:
> If there is code that sets the best solution to a _worse_ solution than we had before, I'd argue that this is the problem we have to fix. 

Actually this is exactly what DefaultCustomPhase do in stepEnded. It updates the best solution only when score is not improved. Is this a bug?

But even if we invert this condition and also disable assigning working solution from best solution in AbstractSolver runPhases for ruin phase only we still cannot "save" the best solution. It will be overwritten by following construction heuristic phase. It by design updates the best solution with working solution in BestSolutionRecaller::processWorkingSolutionDuringConstructionHeuristicsStep. It looks like construction heuristic here is a real problem. We know the order of phases in the AbstractSolver and we could possibly store this information, whether ruin was used before particular construction heuristic and update the best solution accordingly. However I am not sure whether it will not cause some problems in the following local search phase. I quickly commented out this updating best solution part and the local phase claims the solution is uninitialized.

Reply:
The commit that caused this is 258b1c5a0836d7fb684b0ac904860f484015ea7f (which from the commit log, is related to PLANNER-2008).
In particular, the `forceUpdateBestSolution` option was removed, and the condition changed from `if (forceUpdateBestSolution && !bestScoreImproved) {` to `!bestScoreImproved`. Is `forceUpdateBestSolution` changed to always be true or false @triceo ?

Reply:
The current CH is built upon the assumption that it is *the only* CH. Therefore it makes sense for this single CH to set the best solution - after all, it is the first thing that runs, therefore the score *has* to be better.

Obviously, this assumption is no longer valid with the ruin phase. CH can now be run at any time, and therefore it can not assume that it will always provide the best score. I think the CH code needs to be updated, to deal with this new situation.

I don't think LS needs to be updated; LS was always capable of running after another LS, for example. So I think the problem is exclusively in the CH.

Reply:
@Christopher-Chianelli I don't really remember this commit anymore; after all, it's been 3 years already. I'll take a look next week, rebuild my understanding of what goes on in there.